src/iter.rs

```diff
@@ -26,34 +26,32 @@ pub struct PathIter<'a, M: ?Sized, const L: usize, P> {
     separator: &'a str,
 }
 
-impl<'a, M, const L: usize, P> PathIter<'a, M, L, P>
+impl<'a, M, const Y: usize, P> PathIter<'a, M, Y, P>
 where
-    M: Miniconf + ?Sized,
+    M: Miniconf<Y> + ?Sized,
 {
-    pub(crate) fn new(separator: &'a str) -> Result<Self, SliceShort> {
+    pub(crate) fn new(separator: &'a str) -> Self {
         let meta = M::metadata();
-        if L < meta.max_depth {
-            return Err(SliceShort);
-        }
+        assert!(Y == meta.max_depth);
         let mut s = Self::new_unchecked(separator);
         s.count = Some(meta.count);
-        Ok(s)
+        s
     }
 
     pub(crate) fn new_unchecked(separator: &'a str) -> Self {
```

Comment:
I don't see any reason to keep the `new_unchecked()` API any more now that we guarantee, at compile time, the depth equivalency to the type parameter. It seems to me that we should have a single `new()`.

Reply:
Without this we don't get the size hint.

Reply:
Metadata calculation isn't free. It's a recursive traversal of the full tree.